taskrun/TaskManager.py

```diff
@@ -235,112 +245,173 @@ def task_killed(self, task):
-        else:
-          observer.task_failed(task, errors)
+      if task:
+        for observer in self._observers:
+          if task.killed:
+            observer.task_killed(task)
+          else:
+            observer.task_failed(task, errors)
 
       # clean up the task
       self._task_done(task)
 
+  def _kill_running(self, task):
+    """
+    Kills all running tasks except for the specified task.
+
+    Args:
+      task (Task) : A task that should be skipped
+    """
+    # kill all the currently running tasks
+    if not self._killed:
+      self._killed = True
+      for running_task in self._running_tasks:
+        if running_task is not task:
+          running_task.kill()
+
+  def _clear_waiting_and_ready(self):
+    """
+    Clears the waiting and ready tasks lists. Adds the tasks to the filtered
+    list.
+    """
+    # clear out waiting and ready lists
+    self._filter_tasks.extend(self._waiting_tasks)
+    self._waiting_tasks = []
+    for priority in range(self._priority_levels):
+      self._filter_tasks.extend(self._ready_tasks[priority])
+      self._ready_tasks[priority] = []
+
+  def _remove_dependent_tasks(self, task):
+    """
+    Removes the dependent tasks of the specified task. Adds the tasks to the
+    filtered list.
+    """
+    # remove all tasks that depend on this task (BFS)
+    visit = []
+    visit.extend(task.get_dependents())
+    visited = set()
+    while len(visit) > 0:
+      curr = visit.pop()
+      for priority in range(self._priority_levels):
+        assert curr not in self._ready_tasks[priority]
+      assert curr not in self._running_tasks
+      visited.add(curr)
+      for dep in curr.get_dependents():
+        if dep not in visited:
+          visit.append(dep)
+      if curr in self._waiting_tasks:
+        self._waiting_tasks.remove(curr)
+        self._filter_tasks.append(curr)
+
   def _task_done(self, task):
     """
-    This method is called by task_completed and task_failed.
+    This method is called by task_completed and task_error.
     This method removes the task from the running list and gives back all
     resources it consumed.
 
-    WARNING: this method must be called while locked on the condition variable
+    WARNING: this method must be called while locked on the condition variable.
 
     Args:
       task (Task) : the task that finished
     """
     assert self._running is True
 
     # remove task from running lists
-    self._running_tasks.remove(task)
+    if task:
+      self._running_tasks.remove(task)
 
     # give back resources
-    if not task.bypass:
+    if task and not task.bypass:
       if self._resource_manager is not None:
         self._resource_manager.done(task)
 
     # notify waiting thread
     self._condition_variable.notify()
 
+  def _terminate(self):
+    """
+    Executes the procedure for graceful forced shutdown.
+    """
+    self._failure_mode = FailureMode.AGGRESSIVE_FAIL
+    self._task_error(None, None)
```

Comment:
this seems a bit hacky. This feels more natural to me:

_terminate calls _kill_running and _clear_waiting_and_ready
_task_error with FailureMode.AGGRESSIVE_FAIL calls _terminate

then we don't have to worry about the special case of `task` being None.

Reply:
Ok, terminate no longer calls task_error. Terminate just performs what it needs to. task_error does not call terminate (just to keep the code cleaner in the if/elif block).

Reply:
The only function that keeps task=None logic is kill_running, but that happens anyway.